### PR TITLE
FOUNDATIONS: Contribution-ModifyAsync

### LINKED ISSUE #123: FOUNDATIONS: Contribution `ModifyAsync`

### PR DESCRIPTION
**Pull request by SlimAhmad**
Closes #123

### CODE REVIEW

File: GitFyle.Core.Api/Services/Foundations/Contributions/ContributionService.cs [hunk 1]
```diff
@@ -33,5 +33,18 @@ public ValueTask<Contribution> AddContributionAsync(Contribution contribution) =
 
             return await this.storageBroker.InsertContributionAsync(contribution);
         });
+
```
Comment (glhays) [inline GitFyle.Core.Api/Services/Foundations/Contributions/ContributionService.cs:36]: Retrieve methods missing?

File: GitFyle.Core.Api.Tests.Unit/Services/Foundations/Contributions/ContributionServiceTests.Validations.Modify.cs [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsSameAsCreatedDateAndLogItAsync()
+        {
+            //given
+            DateTimeOffset randomDateTimeOffset = GetRandomDateTimeOffset();
```
Comment (glhays) [inline GitFyle.Core.Api.Tests.Unit/Services/Foundations/Contributions/ContributionServiceTests.Validations.Modify.cs:230]:
```suggestion
            // given
```

File: GitFyle.Core.Api.Tests.Unit/Services/Foundations/Contributions/ContributionServiceTests.Validations.Modify.cs [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+
+namespace GitFyle.Core.Api.Tests.Unit.Services.Foundations.Contributions
+{
+    public partial class ContributionServiceTests
+    {
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfContributionIsNullAndLogItAsync()
+        {
+            //given
+            Contribution nullContribution = null;
```
Comment (glhays) [inline GitFyle.Core.Api.Tests.Unit/Services/Foundations/Contributions/ContributionServiceTests.Validations.Modify.cs:21]:
```suggestion
            // given
```

File: GitFyle.Core.Api/Services/Foundations/Contributions/ContributionService.cs [hunk 4]
```diff
@@ -51,5 +51,18 @@ public ValueTask<Contribution> RetrieveContributionByIdAsync(Guid contributionId
 
             return maybeContribution;
         });
+
+        public ValueTask<Contribution> ModifyContributionAsync(Contribution contribution) =>
+        TryCatch(async () =>
+        {
+            await ValidateContributionOnModifyAsync(contribution);
+            Contribution maybeContribution =
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Services/Foundations/Contributions/ContributionService.cs:59]: new line before this line

File: GitFyle.Core.Api/Services/Foundations/Contributions/ContributionService.Validations.cs [hunk 5]
```diff
@@ -16,41 +16,41 @@ private async ValueTask ValidateContributionOnAddAsync(Contribution contribution
...
 
-                (Rule: await IsInvalidLengthAsync(contribution.Title, 255),
-                    Parameter: nameof(Contribution.Title)),
+                (Rule: await IsInvalidLengthAsync(contribution.Title, 255),Parameter: 
+                    nameof(Contribution.Title)),
 
-                (Rule: await IsInvalidLengthAsync(contribution.ExternalId, 255),
-                    Parameter: nameof(Contribution.ExternalId)),
+                (Rule: await IsInvalidLengthAsync(contribution.ExternalId, 255),Parameter: 
+                    nameof(Contribution.ExternalId)),
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Services/Foundations/Contributions/ContributionService.Validations.cs:53]:
```suggestion
            Validate(
                (Rule: await IsInvalidAsync(contribution.Id), Parameter: 
                    nameof(Contribution.Id)),

                (Rule: await IsInvalidAsync(contribution.RepositoryId), Parameter: 
                    nameof(Contribution.RepositoryId)),

                (Rule: await IsInvalidAsync(contribution.ContributorId), Parameter: 
                    nameof(Contribution.ContributorId)),

                (Rule: await IsInvalidAsync(contribution.ContributionTypeId), arameter: 
                    nameof(Contribution.ContributionTypeId)),

                (Rule: await IsInvalidAsync(contribution.ExternalId), Parameter: 
                    nameof(Contribution.ExternalId)),

                (Rule: await IsInvalidAsync(contribution.Title), Parameter: 
                    nameof(Contribution.Title)),

                (Rule: await IsInvalidAsync(contribution.CreatedBy), Parameter: 
                    nameof(Contribution.CreatedBy)),

                (Rule: await IsInvalidAsync(contribution.CreatedDate), Parameter: 
                    nameof(Contribution.CreatedDate)),

                (Rule: await IsInvalidAsync(contribution.UpdatedBy), Parameter: 
                    nameof(Contribution.UpdatedBy)),

                (Rule: await IsInvalidAsync(contribution.UpdatedDate), Parameter: 
                    nameof(Contribution.UpdatedDate)),

                (Rule: await IsInvalidLengthAsync(contribution.Title, 255), Parameter: 
                    nameof(Contribution.Title)),

                (Rule: await IsInvalidLengthAsync(contribution.ExternalId, 255), Parameter: 
                    nameof(Contribution.ExternalId)),
```

File: GitFyle.Core.Api/Services/Foundations/Contributions/ContributionService.Validations.cs [hunk 6]
```diff
@@ -70,6 +70,57 @@ private async ValueTask ValidateContributionOnAddAsync(Contribution contribution
...
+                    nameof(Contribution.UpdatedBy)),
+
+                (Rule: await IsInvalidAsync(contribution.UpdatedDate),Parameter: 
+                    nameof(Contribution.UpdatedDate)),
+
+                (Rule: await IsInvalidLengthAsync(contribution.Title, 255),Parameter: 
+                    nameof(Contribution.Title)),
+
+                (Rule: await IsInvalidLengthAsync(contribution.ExternalId, 255),Parameter: 
+                    nameof(Contribution.ExternalId)),
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Services/Foundations/Contributions/ContributionService.Validations.cs:112]:
```suggestion
                (Rule: await IsInvalidAsync(contribution.RepositoryId), Parameter: 
                    nameof(Contribution.RepositoryId)),

                (Rule: await IsInvalidAsync(contribution.ContributorId), Parameter: 
                    nameof(Contribution.ContributorId)),

                (Rule: await IsInvalidAsync(contribution.ContributionTypeId), Parameter: 
                    nameof(Contribution.ContributionTypeId)),

                (Rule: await IsInvalidAsync(contribution.ExternalId), Parameter: 
                    nameof(Contribution.ExternalId)),

                (Rule: await IsInvalidAsync(contribution.Title), Parameter: 
                    nameof(Contribution.Title)),

                (Rule: await IsInvalidAsync(contribution.CreatedBy), Parameter: 
                    nameof(Contribution.CreatedBy)),

                (Rule: await IsInvalidAsync(contribution.CreatedDate), Parameter: 
                    nameof(Contribution.CreatedDate)),

                (Rule: await IsInvalidAsync(contribution.UpdatedBy), Parameter: 
                    nameof(Contribution.UpdatedBy)),

                (Rule: await IsInvalidAsync(contribution.UpdatedDate), Parameter: 
                    nameof(Contribution.UpdatedDate)),

                (Rule: await IsInvalidLengthAsync(contribution.Title, 255), Parameter: 
                    nameof(Contribution.Title)),

                (Rule: await IsInvalidLengthAsync(contribution.ExternalId, 255), Parameter: 
                    nameof(Contribution.ExternalId)),
```

**Review by cjdutoit — APPROVED** [review]

**Pull request merged**